include/fmt/ranges.h

```diff
@@ -589,34 +589,71 @@ template <typename T, typename Char> struct is_range {
       !std::is_constructible<detail::std_string_view<Char>, T>::value;
 };
 
-template <typename T, typename Char>
+template <typename R, typename Char>
 struct formatter<
-    T, Char,
-    enable_if_t<
-        fmt::is_range<T, Char>::value
+    R, Char,
+    enable_if_t<fmt::is_range<R, Char>::value
 // Workaround a bug in MSVC 2019 and earlier.
 #if !FMT_MSC_VER
-        && (is_formattable<detail::value_type<T>, Char>::value ||
-            detail::has_fallback_formatter<detail::value_type<T>, Char>::value)
+                && (is_formattable<detail::uncvref_type<R>, Char>::value ||
+                    detail::has_fallback_formatter<detail::uncvref_type<R>,
+                                                   Char>::value)
 #endif
-        >> {
+                >> {
+
+  using context = buffer_context<Char>;
+  using mapper = detail::arg_mapper<context>;
+  using element = detail::uncvref_type<R>;
+
+  template <typename T,
+            FMT_ENABLE_IF(has_formatter<remove_cvref_t<T>, context>::value)>
+  static auto map(T&& value) -> T&& {
+    return (T &&) value;
+  }
+  template <typename T,
+            FMT_ENABLE_IF(!has_formatter<remove_cvref_t<T>, context>::value)>
+  static auto map(T&& value) -> decltype(mapper().map((T &&) value)) {
+    return mapper().map((T &&) value);
+  }
+
+  using formatter_type = conditional_t<
+      is_formattable<element, Char>::value,
+      formatter<remove_cvref_t<decltype(map(std::declval<element>()))>, Char>,
+      detail::fallback_formatter<element, Char>>;
+  formatter_type underlying_;
```

Comment:
I more or less copied this from `join`, but I feel like there's probably a more direct way of getting the right `formatter_type`?

Reply:
Not if we want to support implicit conversions. I suggest moving it to a new trait in `detail` and using in both places.